canonical/checkbox
# Fix touchscreen tests on Wayland

## Description

`touchscreen/3-touch-tap` and `touchscreen/4-touch-tap` jobs, which are being used for Desktop certification, are based on `xinput` utility. This does not work well on Wayland sessions (22.04 defaults to Wayland in many cases). However, there are jobs, `touchscreen/evdev/*-touch-tap-*`, based on `python-evdev` that were originally written to be run on systems with no graphical user interface, that work well for this use case.

After re-working the python-evdev based script to work well on 20.04 and 22.04, I changed the related Desktop test plans to point to these jobs instead of the `xinput`-based ones.


## Resolved issues

Fixes [LP#1974197](https://bugs.launchpad.net/plainbox-provider-checkbox/+bug/1974197).


## Documentation

- [ ] Automated tests are included for the changed functionality in this PR. If to be merged without tests, please elaborate why.
  - Python script relies on device hardware and manual intervention to be tested. Cannot be unit-tested.
- [ ] Necessary documentation is provided for the changed functionality in this PR (tests are documentation, too).
  - See commit messages for documentation

## Tests

### How it was tested

Tested on a Dell Inspiron 7370 (laptop with a touchscreen) by running the 20.04 and 22.04 desktop test plan, making sure the correct jobs were displayed in the Touchscreen category, running them and checking they behaved as expected.

I checked the script behavior was as expected on:

- 22.04 Wayland
- 22.04 Xorg
- 22.04 TTY (by entering a TTY with `Ctrl+Alt+F5`)
- 20.04 Xorg
- 20.04 TTY

I also checked the timeout was working as expected in the Checkbox jobs, and would yield a `failed` if no touch event was recorded within 10 seconds.

### Before / After

The `touchscreen-cert-manual` and `after-suspend-touchscreen-cert-manual` nested test plans that were modified are included in `plainbox-provider-certification-client/units/client-cert-desktop-20-04.pxu` and `plainbox-provider-certification-client/units/client-cert-desktop-22-04.pxu`.

With current test plans:

```
u@u-Inspiron-7370:/usr/share$ checkbox-cli list-bootstrapped com.canonical.certification::after-suspend-touchscreen-cert-manual
com.canonical.plainbox::manifest
com.canonical.certification::touchscreen/drag-n-drop
com.canonical.certification::rtc
com.canonical.certification::sleep
com.canonical.certification::suspend/suspend_advanced_auto
com.canonical.certification::after-suspend-touchscreen/drag-n-drop
com.canonical.certification::touchscreen/multitouch-zoom
com.canonical.certification::after-suspend-touchscreen/multitouch-zoom
com.canonical.certification::touchscreen/multitouch-rotate
com.canonical.certification::after-suspend-touchscreen/multitouch-rotate
com.canonical.certification::touchscreen/3-touch-tap
com.canonical.certification::after-suspend-touchscreen/3-touch-tap
com.canonical.certification::touchscreen/4-touch-tap
com.canonical.certification::after-suspend-touchscreen/4-touch-tap
```

With updated test plans on my Inspiron test device:

```
u@u-Inspiron-7370:/usr/share$ checkbox-cli list-bootstrapped com.canonical.certification::touchscreen-cert-manual
com.canonical.plainbox::manifest
com.canonical.certification::touchscreen/drag-n-drop
com.canonical.certification::touchscreen/multitouch-zoom
com.canonical.certification::touchscreen/multitouch-rotate
com.canonical.certification::touchscreen/evdev/single-touch-tap-ELAN2097_00_04F3_2365
com.canonical.certification::touchscreen/evdev/2-touch-tap-ELAN2097_00_04F3_2365
com.canonical.certification::touchscreen/evdev/3-touch-tap-ELAN2097_00_04F3_2365
com.canonical.certification::touchscreen/evdev/4-touch-tap-ELAN2097_00_04F3_2365


u@u-Inspiron-7370:/usr/share$ checkbox-cli list-bootstrapped com.canonical.certification::after-suspend-touchscreen-cert-manual
com.canonical.certification::rtc
com.canonical.certification::sleep
com.canonical.certification::suspend/suspend_advanced_auto
com.canonical.plainbox::manifest
com.canonical.certification::touchscreen/drag-n-drop
com.canonical.certification::after-suspend-touchscreen/drag-n-drop
com.canonical.certification::touchscreen/multitouch-zoom
com.canonical.certification::after-suspend-touchscreen/multitouch-zoom
com.canonical.certification::touchscreen/multitouch-rotate
com.canonical.certification::after-suspend-touchscreen/multitouch-rotate
com.canonical.certification::touchscreen/evdev/single-touch-tap-ELAN2097_00_04F3_2365
com.canonical.certification::touchscreen/evdev/2-touch-tap-ELAN2097_00_04F3_2365
com.canonical.certification::touchscreen/evdev/3-touch-tap-ELAN2097_00_04F3_2365
com.canonical.certification::after-suspend-touchscreen/evdev/3-touch-tap-ELAN2097_00_04F3_2365
com.canonical.certification::touchscreen/evdev/4-touch-tap-ELAN2097_00_04F3_2365
com.canonical.certification::after-suspend-touchscreen/evdev/4-touch-tap-ELAN2097_00_04F3_2365
```

### How to test

1. After applying the patch, run checkbox, select 22.04 (or 20.04) desktop test plan
2. Make sure the right jobs are available in the Touchscreen section. If you display the job ids (by pressing `i` key) you should see something like:
  - `touchscreen/evdev/3-touch-tap-ELAN2097_00_04F3_2365`
  - `after-suspend-touchscreen/evdev/3-touch-tap-ELAN2097_00_04F3_2365`
  - `touchscreen/evdev/4-touch-tap-ELAN2097_00_04F3_2365`
  - `after-suspend-touchscreen/evdev/4-touch-tap-ELAN2097_00_04F3_2365`
3. Run the aforementioned jobs and make sure they behave as expected.
